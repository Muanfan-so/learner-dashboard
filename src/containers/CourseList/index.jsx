import React from 'react';
import PropTypes from 'prop-types';

import CourseCard from 'containers/CourseCard';
import SelectSession from 'containers/SelectSession';

export const CourseList = ({ courseListData }) => (
  <div className="d-flex flex-column flex-grow-1">
    {courseListData.map((courseNumber) => (
      <CourseCard key={courseNumber} courseNumber={courseNumber} />
    ))}
    <SelectSession />
  </div>
);

CourseList.propTypes = {
  courseListData: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default CourseList;
