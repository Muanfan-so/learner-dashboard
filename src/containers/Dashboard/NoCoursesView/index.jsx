import React from 'react';
import classNames from 'classnames';
import { FormattedMessage } from '@edx/frontend-platform/i18n';
import { Button, Image } from '@edx/paragon';
import { Search } from '@edx/paragon/icons';

import emptyCourseSVG from 'assets/empty-course.svg';

import { hooks as appHooks } from 'data/redux';
import messages from './messages';

import './index.scss';

export const NoCoursesView = () => {
  const { courseSearchUrl } = appHooks.usePlatformSettingsData();
  return (
    <div
      className={classNames(
        'd-flex align-items-center justify-content-center empty-course-hero',
      )}
    >
      <Image src={emptyCourseSVG} alt="empty course banner" />
      <h1>
        <FormattedMessage {...messages.lookingForChallengePrompt} />
      </h1>
      <p>
        <FormattedMessage {...messages.exploreCoursesPrompt} />
      </p>
      <Button
        variant="brand"
        as="a"
        href={courseSearchUrl}
        iconBefore={Search}
      >
        <FormattedMessage {...messages.exploreCoursesButton} />
      </Button>
    </div>
  );
};

export default NoCoursesView;
