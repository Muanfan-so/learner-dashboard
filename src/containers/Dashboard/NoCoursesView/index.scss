@import "@edx/paragon/scss/core/core";

.empty-course-hero {
  flex-direction: column;
  padding-bottom: map-get($spacers, 5);
  padding-top: map-get($spacers, 5);
  height: 100%;

  & > * {
    margin-top: map-get($spacers, 3);
    margin-bottom: map-get($spacers, 3);
  }
}

.empty-course-border {
  border: 2px solid $light-400;
}
