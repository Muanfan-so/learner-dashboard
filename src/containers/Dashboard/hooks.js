import React from 'react';
import { useWindowSize, breakpoints } from '@edx/paragon';
import { useDispatch } from 'react-redux';
import { thunkActions } from 'data/redux';

export const useIsDashboardCollapsed = () => {
  const { width } = useWindowSize();
  return width < breakpoints.large.maxWidth;
};

export const useInitializeDashboard = () => {
  const dispatch = useDispatch();
  React.useEffect(
    () => { dispatch(thunkActions.app.initialize()); },
    [dispatch],
  );
};

export default {
  useIsDashboardCollapsed,
  useInitializeDashboard,
};
