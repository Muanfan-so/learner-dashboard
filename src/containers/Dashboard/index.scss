@import "@edx/paragon/scss/core/core";

.course-list-column {
  padding: 0 map-get($spacers, 4);
  margin: auto;
}

.sidebar-column {
  padding: 0 map-get($spacers, 3) 0 map-get($spacers, 1);
}

@include media-breakpoint-down(xl) {
  .sidebar-column {
    padding: 0 map-get($spacers, 3) !important;
  }
}