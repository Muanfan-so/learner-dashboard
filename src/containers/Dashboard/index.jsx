import React from 'react';
import { useDispatch } from 'react-redux';
import { Container, Col, Row } from '@edx/paragon';

import {
  thunkActions,
  hooks as appHooks,
} from 'data/redux';
import { RequestKeys } from 'data/constants/requests';

import CourseList from 'containers/CourseList';
import WidgetSidebar from 'containers/WidgetSidebar';
import EmptyCourse from 'containers/EmptyCourse';
import SelectSessionModal from 'containers/SelectSessionModal';
import EnterpriseDashboardModal from 'containers/EnterpriseDashboardModal';

import './index.scss';

export const Dashboard = () => {
  const dispatch = useDispatch();
  React.useEffect(
    () => { dispatch(thunkActions.app.initialize()); },
    [dispatch],
  );

  const hasCourses = appHooks.useHasCourses();
  const hasAvailableDashboards = appHooks.useHasAvailableDashboards();
  const showSelectSessionModal = appHooks.useShowSelectSessionModal();
  const initIsPending = appHooks.useIsPendingRequest(RequestKeys.initialize);

  return (
    <div id="dashboard-container" className="d-flex flex-column p-2 mt-3">
      {hasAvailableDashboards && <EnterpriseDashboardModal />}
      {initIsPending || (!initIsPending && hasCourses) ? (
        <Container fluid size="xl">
          <Row>
            <Col
              xs={{ span: 12, offset: 0 }}
              sm={{ span: 8, offset: 2 }}
              md={{ span: 12, offset: 0 }}
              lg={{ span: 10, offset: 1 }}
              xl={{ span: 8, offset: 0 }}
              className="p-0 px-4"
            >
              {showSelectSessionModal && (<SelectSessionModal />)}
              <CourseList />
            </Col>
            <Col md={12} xl={4} className="p-0 pr-4 pl-1">
              <WidgetSidebar />
            </Col>
          </Row>
        </Container>
      ) : (<EmptyCourse />)}
    </div>
  );
};

export default Dashboard;
