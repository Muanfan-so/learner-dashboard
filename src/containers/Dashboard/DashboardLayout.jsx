import React from 'react';
import PropTypes from 'prop-types';

import { Container, Col, Row } from '@edx/paragon';

import hooks from './hooks';

export const columnConfig = {
  courseList: {
    lg: { span: 12, offset: 0 },
    xl: { span: 8, offset: 0 },
  },
  sidebar: {
    lg: { span: 12, offset: 0 },
    xl: { span: 4, offset: 0 },
  },
};

export const DashboardLayout = ({ bottomWidgetContainer, children, sidebar }) => {
  const isCollapsed = hooks.useIsDashboardCollapsed();

  return (
    <Container fluid size="xl">
      <Row>
        <Col {...columnConfig.courseList} className="course-list-column">
          {children}
        </Col>
        <Col {...columnConfig.sidebar} className="sidebar-column">
          {!isCollapsed && (<h2 className="course-list-title">&nbsp;</h2>)}
          {sidebar}
        </Col>
      </Row>
      <Row>
        <Col>
          {bottomWidgetContainer}
        </Col>
      </Row>
    </Container>
  );
};
DashboardLayout.propTypes = {
  bottomWidgetContainer: PropTypes.node.isRequired,
  children: PropTypes.node.isRequired,
  sidebar: PropTypes.node.isRequired,
};

export default DashboardLayout;
