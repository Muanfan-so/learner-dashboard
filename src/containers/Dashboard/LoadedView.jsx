import React from 'react';
import classNames from 'classnames';
import {
  Container,
  Col,
  Row,
} from '@edx/paragon';

import { hooks as appHooks } from 'data/redux';
import { RequestKeys } from 'data/constants/requests';
import CourseList from 'containers/CourseList';
import RecommendationsPanel, { LoadingView as RecommendationsLoadingView } from 'containers/RecommendationsPanel';
import WidgetSidebar from 'containers/WidgetSidebar';
import hooks from './hooks';

const commonColumnConfig = {
  sm: { span: 12, offset: 0 },
  md: { span: 10, offset: 1 },
  lg: { span: 12, offset: 0 },
};

export const columnConfig = {
  courseList: {
    ...commonColumnConfig,
    xl: { span: 8, offset: 0 },
  },
  sidebar: {
    ...commonColumnConfig,
    xl: { span: 4, offset: 0 },
  },
};

export const LoadedView = () => {
  const isCollapsed = hooks.useIsDashboardCollapsed();
  const recommendedCourses = appHooks.useRecommendedCoursesData();
  const hasRecommendedCourses = recommendedCourses.courses.length > 0;
  const isRecommendationsPending = appHooks.useRequestIsPending(RequestKeys.recommendedCourses);

  return (
    <Container fluid size="xl">
      <Row>
        <Col {...columnConfig.courseList} className="course-list-column">
          <CourseList />
        </Col>
        <Col {...columnConfig.sidebar} className="sidebar-column">
          {!isCollapsed && (<h2 className="course-list-title">&nbsp;</h2>)}
          <WidgetSidebar />
        </Col>
      </Row>
    </Container>
  );
};

export default LoadedView;
