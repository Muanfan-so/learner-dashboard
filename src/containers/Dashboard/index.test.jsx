import { shallow } from 'enzyme';

import { hooks } from 'data/redux';

import CourseList from 'containers/CourseList';
import WidgetSidebar from 'containers/WidgetSidebar';
import EmptyCourse from 'containers/EmptyCourse';
import SelectSessionModal from 'containers/SelectSessionModal';
import EnterpriseDashboardModal from 'containers/EnterpriseDashboardModal';

import Dashboard from '.';

jest.mock('data/redux', () => ({
  thunkActions: {
    app: {
      initialize: jest.fn(),
    },
  },
  hooks: {
    useHasCourses: jest.fn(),
    useHasAvailableDashboards: jest.fn(),
    useShowSelectSessionModal: jest.fn(),
  },
}));

jest.mock('containers/CourseList', () => 'CourseList');
jest.mock('containers/WidgetSidebar', () => 'WidgetSidebar');
jest.mock('containers/EmptyCourse', () => 'EmptyCourse');
jest.mock('containers/SelectSessionModal', () => 'SelectSessionModal');
jest.mock('containers/EnterpriseDashboardModal', () => 'EnterpriseDashboardModal');

describe('Dashboard', () => {
  const createWrapper = ({
    hasCourses,
    hasAvailableDashboards,
    showSelectSessionModal,
  }) => {
    hooks.useHasCourses.mockReturnValueOnce(hasCourses);
    hooks.useHasAvailableDashboards.mockReturnValueOnce(hasAvailableDashboards);
    hooks.useShowSelectSessionModal.mockReturnValueOnce(showSelectSessionModal);
    return shallow(<Dashboard />);
  };

  describe('snapshots', () => {
    test('there are courses', () => {
      const wrapper = createWrapper({
        hasCourses: true,
        hasAvailableDashboards: false,
        showSelectSessionModal: false,
      });
      expect(wrapper).toMatchSnapshot();
    });

    test('there are no courses', () => {
      const wrapper = createWrapper({
        hasCourses: false,
        hasAvailableDashboards: false,
        showSelectSessionModal: false,
      });
      expect(wrapper).toMatchSnapshot();
    });

    test('there are available dashboards', () => {
      const wrapper = createWrapper({
        hasCourses: false,
        hasAvailableDashboards: true,
        showSelectSessionModal: false,
      });
      expect(wrapper).toMatchSnapshot();
    });

    test('there is a select session modal', () => {
      const wrapper = createWrapper({
        hasCourses: false,
        hasAvailableDashboards: false,
        showSelectSessionModal: true,
      });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('initializes the app without courses', () => {
      const wrapper = createWrapper({
        hasCourses: false,
        hasAvailableDashboards: false,
        showSelectSessionModal: false,
      });
      expect(wrapper.find(EmptyCourse).length).toEqual(1);
      expect(wrapper.find(CourseList).length).toEqual(0);
      expect(wrapper.find(WidgetSidebar).length).toEqual(0);
      expect(wrapper.find(SelectSessionModal).length).toEqual(0);
      expect(wrapper.find(EnterpriseDashboardModal).length).toEqual(0);
    });
    it('initializes the app with courses, dashboard and select', () => {
      const wrapper = createWrapper({
        hasCourses: true,
        hasAvailableDashboards: true,
        showSelectSessionModal: true,
      });
      expect(wrapper.find(EmptyCourse).length).toEqual(0);
      expect(wrapper.find(CourseList).length).toEqual(1);
      expect(wrapper.find(WidgetSidebar).length).toEqual(1);
      expect(wrapper.find(SelectSessionModal).length).toEqual(1);
      expect(wrapper.find(EnterpriseDashboardModal).length).toEqual(1);
    });
    it('initializes the app with courses, dashboard and no select', () => {
      const wrapper = createWrapper({
        hasCourses: true,
        hasAvailableDashboards: true,
        showSelectSessionModal: false,
      });
      expect(wrapper.find(EmptyCourse).length).toEqual(0);
      expect(wrapper.find(CourseList).length).toEqual(1);
      expect(wrapper.find(WidgetSidebar).length).toEqual(1);
      expect(wrapper.find(SelectSessionModal).length).toEqual(0);
      expect(wrapper.find(EnterpriseDashboardModal).length).toEqual(1);
    });
  });
});
