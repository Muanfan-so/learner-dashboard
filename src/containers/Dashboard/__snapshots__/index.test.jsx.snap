// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard snapshots there are available dashboards 1`] = `
<div
  className="d-flex flex-column p-2 mt-3"
  id="dashboard-container"
>
  <EnterpriseDashboardModal />
  <EmptyCourse />
</div>
`;

exports[`Dashboard snapshots there are courses, or they are still loading 1`] = `
<div
  className="d-flex flex-column p-2 mt-3"
  id="dashboard-container"
>
  <Container
    fluid={true}
    size="xl"
  >
    <Row>
      <Col
        className="p-0 px-4"
        lg={
          Object {
            "offset": 1,
            "span": 10,
          }
        }
        md={
          Object {
            "offset": 0,
            "span": 12,
          }
        }
        sm={
          Object {
            "offset": 2,
            "span": 8,
          }
        }
        xl={
          Object {
            "offset": 0,
            "span": 8,
          }
        }
        xs={
          Object {
            "offset": 0,
            "span": 12,
          }
        }
      >
        <CourseList />
      </Col>
      <Col
        className="p-0 pr-4 pl-1"
        md={12}
        xl={4}
      >
        <WidgetSidebar />
      </Col>
    </Row>
  </Container>
</div>
`;

exports[`Dashboard snapshots there are no courses 1`] = `
<div
  className="d-flex flex-column p-2 mt-3"
  id="dashboard-container"
>
  <EmptyCourse />
</div>
`;

exports[`Dashboard snapshots there is a select session modal 1`] = `
<div
  className="d-flex flex-column p-2 mt-3"
  id="dashboard-container"
>
  <EmptyCourse />
</div>
`;
